tools/mksurfdata_esmf/src/mksoiltexMod.F90

```diff
@@ -35,6 +35,81 @@ module mksoiltexMod
 contains
 !=================================================================================
 
+  subroutine mksoiltex_i_to_o(no, lookup_index, n_scid, nlay, n_mapunits, name, val_neg_4, val_neg_other, data_i, data_o, data_modifier, organic_o)
+    !
+    ! Arguments
+    integer, intent(in) :: no
+    integer, intent(in) :: lookup_index
+    integer, intent(in) :: n_scid, nlay, n_mapunits
+    character(*), intent(in) :: name
+    real(r4), intent(in) :: val_neg_4  ! Fallback value if read-in value is -4
+    real(r4), intent(in) :: val_neg_other  ! Fallback value if read-in value is negative but not -4
+    real(r4), intent(in) :: data_i(:,:,:)
+    real(r4), intent(out) :: data_o(:,:)
+    real(r4), optional, intent(in) :: data_modifier(:,:,:)
+    real(r4), optional, intent(out) :: organic_o(:,:)
+    !
+    ! Local variables
+    integer :: l
+    integer :: ier
+    logical :: is_sand_dune
+    real(r4), allocatable  :: organic_i(:,:,:)
+
+    ! Apply data_modifer, if needed (organic_o OPTION 2)
+    if (present(data_modifier)) then
+       if (.not. present(organic_o)) then
+          call shr_sys_abort('mksoiltex_i_to_o: data_modifier not needed if not providing organic_o')
+       else if (trim(name) /= 'orgc') then
+          call shr_sys_abort('mksoiltex_i_to_o: organic_o should only be provided along with orgc')
+       end if
+       allocate(organic_i(nlay,n_scid,n_mapunits), stat=ier)
+       if (ier/=0) call shr_sys_abort()
+       organic_i = data_i * data_modifier
+    else if (present(organic_o)) then
+       call shr_sys_abort('mksoiltex_i_to_o: If providing organic_o, also provide data_modifier')
+    end if
+
+    ! Fill first layer of output array with first positive value on SCID dim of input array
+    data_o(no,1) = data_i(1,1,lookup_index)
+    if (data_o(no,1) < 0._r4) then
+       do l = 2,n_scid
+          if (data_i(1,l,lookup_index) >= 0._r4) then
+             data_o(no,1) = data_i(1,l,lookup_index)
+             exit
+          end if
+       end do
+    end if
+
+    ! Handle cases where no positive value was found
+    if (data_o(no,1) < 0._r4) then
+       is_sand_dune = int(data_o(no,1)) == -4
+       if (is_sand_dune) then
+          data_o(no,:) = val_neg_4
+       else
+          data_o(no,:) = val_neg_other
+       end if
+    end if
+
+    ! Ensure negative values are handled
+    ! TODO: Remove as unnecessary?
+    ! TODO: Also handle organic_o? Not handled in original
+    if (data_o(no,1) < 0._r4 .and. trim(name) /= "sand") then
+       write(6,*)'ERROR: at no, lookup_index = ',no,lookup_index
+       call shr_sys_abort('could not find a value >= 0 for '//name)
+    end if
```

Comment:
@slevis-lmwg @ekluzek The last question I have pertains to this "Ensure negative values are handled" block. In the original code, sand and `organic_o` weren't checked for negative values. Should they be?

Reply:
Checking that data read in is in the range it's expected to be in, is good standard error checking that's worth including. Strange things can happen to input data files and if they have bad values, you want to know as soon as possible. This is something that I think we should always have in place. Otherwise, it's possible you could run with a bad input file, and not know about any problems until far down the road when you've run simulations and done analysis. These are much more expensive type of problems to have, where it silently does the wrong thing.

So I would remove the "Remove as unnecessary?" comment. And also add handling for organic_o.